File: src/main/scala/com/klymenko/pawn/scala/UniqueMovementAlgorithm.scala
package com.klymenko.pawn.scala

import java.util

import scala.collection.mutable.ListBuffer

/**
  * Created by Unicorn on 26.03.2017.
  */
object UniqueMovementAlgorithm {
  def findPath(board: Board, currentCell: GameCell) {
    if(board.shouldBeVisited(currentCell)) {
      board.visit(currentCell)
      for(availableCell <- getAvailableCells(currentCell, board)) {
        if(board.shouldBeVisited(availableCell)) {
          findPath(board, availableCell)
        }
      }
    }
  }

  def getAvailableCells(currentCell : GameCell, board : Board) = {
    val availableCells = ListBuffer[GameCell]();
    if(currentCell.y - board.straightStep >= 0) availableCells += new GameCell(currentCell.x, currentCell.y - board.straightStep);
    if(currentCell.y + board.straightStep < board.boardSize) availableCells += new GameCell(currentCell.x, currentCell.y + board.straightStep);
    if(currentCell.x - board.straightStep >= 0) availableCells += new GameCell(currentCell.x - board.straightStep, currentCell.y);
    if(currentCell.x + board.straightStep < board.boardSize) availableCells += new GameCell(currentCell.x + board.straightStep, currentCell.y);
    if(currentCell.x + board.diagonalStep < board.boardSize && currentCell.y - board.diagonalStep >= 0) availableCells += new GameCell(currentCell.x + board.diagonalStep, currentCell.y - board.diagonalStep)
    if(currentCell.x - board.diagonalStep >= 0 && currentCell.y - board.diagonalStep >= 0) availableCells += new GameCell(currentCell.x - board.diagonalStep, currentCell.y - board.diagonalStep)
    if(currentCell.x + board.diagonalStep < board.boardSize && currentCell.y + board.diagonalStep < board.boardSize) availableCells += new GameCell(currentCell.x + board.diagonalStep, currentCell.y + board.diagonalStep)
    if(currentCell.x - board.diagonalStep >= 0 && currentCell.y + board.diagonalStep < board.boardSize) availableCells += new GameCell(currentCell.x - board.diagonalStep, currentCell.y + board.diagonalStep)
    availableCells;
  }
}

File: src/main/java/com/klymenko/pawn/UniqueMovementAlgorithm.java
package com.klymenko.pawn;

import java.util.LinkedList;
import java.util.List;

public class UniqueMovementAlgorithm {
	
	public static void findPath(Board board, GameCell currentCell) {
		if(board.shouldBeVisited(currentCell)) {
			board.visit(currentCell);
			for(GameCell availableCell : getAvailableCells(currentCell, board)) {
				if(board.shouldBeVisited(availableCell)) {
					findPath(board, availableCell);
				}
			}
		}
	}
	
	private static List<GameCell> getAvailableCells(GameCell currentCell, Board board) {
		List<GameCell> availableCells = new LinkedList<>();
		addNorthCell(availableCells, currentCell, board);
		addSouthCell(availableCells, currentCell, board);
		addWestCell(availableCells, currentCell, board);
		addEastCell(availableCells, currentCell, board);
		addNorthWestCell(availableCells, currentCell, board);
		addNorthEastCell(availableCells, currentCell, board);
		addSouthWestCell(availableCells, currentCell, board);
		addSouthEastCell(availableCells, currentCell, board);
		return availableCells;
	}
	private static void addNorthCell(List<GameCell> availableCells, GameCell currentCell, Board board) {
		if(currentCell.getY() - board.getStraightStep() >= 0) {
			availableCells.add(new GameCell(currentCell.getX(), currentCell.getY() - board.getStraightStep()));
		}
	}
	private static void addSouthCell(List<GameCell> availableCells, GameCell currentCell, Board board) {
		if(currentCell.getY() + board.getStraightStep() < board.getBoardSize()) {
			availableCells.add(new GameCell(currentCell.getX(), currentCell.getY() + board.getStraightStep()));
		}
	}
	private static void addWestCell(List<GameCell> availableCells, GameCell currentCell, Board board) {
		if(currentCell.getX() - board.getStraightStep() >= 0) {
			availableCells.add(new GameCell(currentCell.getX() - board.getStraightStep(), currentCell.getY()));
		}
	}
	private static void addEastCell(List<GameCell> availableCells, GameCell currentCell, Board board) {
		if(currentCell.getX() + board.getStraightStep() < board.getBoardSize()) {
			availableCells.add(new GameCell(currentCell.getX() + board.getStraightStep(), currentCell.getY()));
		}
	}
	private static void addNorthWestCell(List<GameCell> availableCells, GameCell currentCell, Board board) {
		if(currentCell.getX() + board.getDiagonalStep() < board.getBoardSize() &&
		   currentCell.getY() - board.getDiagonalStep() >= 0) {
			availableCells.add(new GameCell(currentCell.getX() + board.getDiagonalStep(), currentCell.getY() - board.getDiagonalStep()));
		}
	}
	private static void addNorthEastCell(List<GameCell> availableCells, GameCell currentCell, Board board) {
		if(currentCell.getX() - board.getDiagonalStep() >= 0 &&
		   currentCell.getY() - board.getDiagonalStep() >= 0) {
			availableCells.add(new GameCell(currentCell.getX() - board.getDiagonalStep(), currentCell.getY() - board.getDiagonalStep()));
		}
	}
	private static void addSouthWestCell(List<GameCell> availableCells, GameCell currentCell, Board board) {
		if(currentCell.getX() + board.getDiagonalStep() < board.getBoardSize() &&
		   currentCell.getY() + board.getDiagonalStep() < board.getBoardSize()) {
			availableCells.add(new GameCell(currentCell.getX() + board.getDiagonalStep(), currentCell.getY() + board.getDiagonalStep()));
		}
	}
	private static void addSouthEastCell(List<GameCell> availableCells, GameCell currentCell, Board board) {
		if(currentCell.getX() - board.getDiagonalStep() >= 0 &&
		   currentCell.getY() + board.getDiagonalStep() < board.getBoardSize()) {
			availableCells.add(new GameCell(currentCell.getX() - board.getDiagonalStep(), currentCell.getY() + board.getDiagonalStep()));
		}
	}
}
